File: code/code105.py
from typing import List,Optional
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        self.curr = 0
        if len(preorder) == 0:
            return None
        n = len(preorder)
        def createTree(left:int,right:int) -> Optional[TreeNode]:
            if left > right:
                return None
            root = TreeNode(preorder[self.curr])
            self.curr += 1
            index = 0
            for index in range(left,right+1):
                if inorder[index] == root.val:
                    break
            root.left = createTree(left,index-1)
            root.right = createTree(index+1,right)
            return root
        tree = createTree(0,n-1)
        return tree
    def buildTree_pro(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        '''
        用map记录下标
        '''
        self.curr = 0
        if not preorder:
            return None
        n = len(preorder)
        hash_map = {val:idx for idx,val in enumerate(inorder)}
        def createTree(left:int,right:int) -> Optional[TreeNode]:
            if left > right:
                return None
            root = TreeNode(preorder[self.curr])
            self.curr += 1
            index = hash_map[root.val]
            root.left = createTree(left,index-1)
            root.right = createTree(index+1,right)
            return root
        tree = createTree(0,n-1)
        return tree
if __name__ == "__main__":
    soultion = Solution()
    preorder = [1,2]
    inorder = [1,2]
    tree = soultion.buildTree(preorder,inorder)

File: code/code105.cpp
#include<bits/stdc++.h>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution {
    int dx = 0;
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return buideTreeWithInorder(preorder,inorder,0,preorder.size()-1);
    }
    TreeNode* buideTreeWithInorder(vector<int>& preorder, vector<int>& inorder
        ,int left,int right) {
        if (left == right) {
            return new TreeNode(preorder[dx++]);
        }
        if (left > right) {
            return nullptr;
        }
        int tNodeVal = preorder[dx++];
        TreeNode* tree = new TreeNode(tNodeVal);
        int index = left;
        for (int i = left;i <= right;i++) {
            if (inorder[i] == tNodeVal) {
                index = i;
                break;
            }
        }
        tree->left = buideTreeWithInorder(preorder,inorder,left,index-1);
        tree->right = buideTreeWithInorder(preorder,inorder,index+1,right);
        return tree;
    }
};
int main() {
    vector<int> preo = {3,9,20,15,7};
    vector<int> ino = {9,3,15,20,7};
    Solution solution;
    TreeNode* tree = solution.buildTree(preo,ino);
}